cyclonedds/idl/_main.py

```diff
@@ -209,45 +209,14 @@ def _deserialize(self, data, has_header=True, use_version_2: bool = None, deseri
 
         return machine.deserialize(buffer, deserialize_kind=deserialize_kind)
 
-    def deserialize(self, data, has_header=True, use_version_2: bool = None) -> object:
-        return self._deserialize(data, has_header, use_version_2, DeserializeKind.DataSample)
-
     def deserialize_key(self, data, has_header=True, use_version_2: bool = None) -> object:
-        return self._deserialize(data, has_header, use_version_2, DeserializeKind.KeySample)
-
-    def _serialize_key(self, object, use_version_2: bool = None, endianness: Endianness = Endianness.Little, serialize_kind: SerializeKind = SerializeKind.KeyNormalized) -> bytes:
-        if not self._populated:
-            self.populate()
-
-        if self.version_support.SupportsBasic & self.version_support:
-            use_version_2 = False if use_version_2 is None else use_version_2
-        else:
-            # version 0 not supported
-            if use_version_2 is not None and not use_version_2:
-                raise Exception("Cannot encode this type with version 0, contains xcdrv2-type structures")
-            use_version_2 = True
-
-        if self.keyless:
```

Comment:
I concluded that it was really just an optimisation: if there are no keys, then `self.v{2,0}_machine.serialize` skips all the fields and yields an empty output. I figured the optimisation wasn't really worth the bother, because one doesn't really serialise keys all that often anyway.

I'll have another look, see if I still agree with my original analysis.